### PR TITLE
basedpyright: init at 1.12.3

### LINKED ISSUE #306129: Package request: python3packages.basedpyright
**Project description**

Basedpyright is a fork of [pyright](https://github.com/microsoft/pyright) with various type checking improvements, improved vscode support and pylance features built into the language server.

**Metadata**

* homepage URL: https://detachhead.github.io/basedpyright/#/
* source URL: https://github.com/detachhead/basedpyright
* license: mit
* platforms: unix, linux, darwin

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by kiike**
## Description of changes

This PR introduces a package for basedpyright, a fork of Microsoft's pyright that introduces some changes such as introduction of some improved unreachable code heuristics, banning of types, and adding features that would be only available with Pylance.

Closes #306129 

## Things done

Since this packages a fork, I used the existing pyright `package.nix` by @kalekseev  (thanks!) and changed the variables accordingly. I changed the executable paths so hopefully both pyright and basedpyright can coexist side-by-side.

I checked the functionality of the binaries using the helix editor. Instructions for testing this kind of setup will be in a comment below.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by kiike** [comment]
To test:
1. I have started a nix-shell with `nix shell .#basedpyright`
2. included the basedpyright helix config in my home-manager config:

```
programs.helix.languages."language-server" = {
  basedpyright = {
    command = "basedpyright-langserver";
    args = [ "--stdio" ];
    config = { };
    };
};
```

3. Included the config in a Python project:
```
cat .helix/languages.toml
[[language]]
name = "python"
language-servers = ["basedpyright"]
```

4. Tested whether the configuration is fine:
```
hx --health python
Configured language servers:
  ✓ basedpyright-langserver: /nix/store/zrlvsrsflb7d8yxww42fgs4g12vcr47n-basedpyright-1.11.0/bin/basedpyright-langserver
Configured debug adapter: python3
Binary for debug adapter: /nix/store/3v2ch16fkl50i85n05h5ckss8pxx6836-python3-3.11.8/bin/python3
Configured formatter: None
Highlight queries: ✓
Textobject queries: ✓
Indent queries: ✓
```

image

**Comment by getchoo** [comment]
Result of `nixpkgs-review pr 308503` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>basedpyright</li>
  </ul>
</details>

**Comment by kiike** [comment]
Hi @kalekseev! Seeing that you reviewed and developed the pyright package, would you like to review this?

**Comment by kalekseev** [comment]
> Hi @kalekseev! Seeing that you reviewed and developed the pyright package, would you like to review this?

Looks good to me but I don't have rights to merge.

**Review by kalekseev — APPROVED** [review]

**Comment by pluiedev** [comment]
Result of `nixpkgs-review pr 308503` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>basedpyright</li>
  </ul>
</details>

**Review by pluiedev — APPROVED** [review]

**Comment by kiike** [comment]
Thank you both for your approvals!

**Comment by kiike** [comment]
Updated to latest version released today.

**Review by pluiedev — APPROVED** [review]
Works great!

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 308503` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>basedpyright</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 308503` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>basedpyright</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 308503` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>basedpyright</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 308503` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>basedpyright</li>
  </ul>
</details>

**Comment by kiike** [comment]
Updated to 1.12.3 released yesterday:

Changelog:

> update upstream pyright version to 1.1.364 by @DetachHead in https://github.com/DetachHead/basedpyright/pull/375
>     fix Any being coloured as a variable instead of a type by @DetachHead in https://github.com/DetachHead/basedpyright/pull/373
>     documentation enhancements
>         new URL: https://docs.basedpyright.com/
>         explain how to use the CLI & language server from pypi and make it clear that many of the IDE plugins require them to be installed by @DetachHead in https://github.com/DetachHead/basedpyright/pull/366
>         Improve Neovim installation instructions by @baco in https://github.com/DetachHead/basedpyright/pull/368

Full Changelog: https://github.com/DetachHead/basedpyright/compare/v1.12.2...v1.12.3

Testing it in a shell with `nix shell .#basedpyright -L`.

Copy & paste command:

```bash
cat << EOF > test.py; basedpyright --version; basedpyright test.py
import sys

if sys.platform == "win32":
    a = "a" + 1
EOF
```
Output:
```
basedpyright 1.12.3
based on pyright 1.1.364
/home/kiike/projects/nixpkgs/test.py
  /home/kiike/projects/nixpkgs/test.py:4:5 - error: Type of "a" is unknown (reportUnknownVariableType)
  /home/kiike/projects/nixpkgs/test.py:4:9 - error: Operator "+" not supported for types "Literal['a']" and "Literal[1]" (reportOperatorIssue)
2 errors, 0 warnings, 0 notes
```

**Review by GaetanLepage — APPROVED** [review]
LGTM !

**Comment by GaetanLepage** [comment]
I would squash all the commits into one: `basedpyright: init at 1.12.3` ?

File: pkgs/by-name/ba/basedpyright/package.nix [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  '';
+
+  postInstall = ''
+    mv "$out/bin/pyright" "$out/bin/basedpyright"
+    mv "$out/bin/pyright-langserver" "$out/bin/basedpyright-langserver"
+  '';
+
+  dontNpmBuild = true;
+
+  passthru.updateScript = nix-update-script { };
```
Comment (natsukium) [inline pkgs/by-name/ba/basedpyright/package.nix:79]: Does this update script work properly?
The original pyright has its own script.

**Review by natsukium — APPROVED** [review]

**Comment by kiike** [comment]
> I would squash all the commits into one: `basedpyright: init at 1.12.3`.

Done!

File: pkgs/by-name/ba/basedpyright/package.nix (hunk 1 above)
Comment (kiike) [reply pkgs/by-name/ba/basedpyright/package.nix:79]: The original script didn't work for me. I tried nix-update-script and it worked.

**Review by natsukium — APPROVED** [review]
LGTM, thank you!

**Pull request merged**